# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [296]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
from getpass import getpass
import mysql.connector


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [299]:
password = getpass.getpass("Enter MySQL password: ")

Enter MySQL password:  ········


In [438]:
# Establish a connection to mysql
connection = mysql.connector.connect(
    host="127.0.0.1",  
    user="root",       
    password=password, 
    database="austin_weather"  
)

In [440]:
# Check connection 
if connection.is_connected():
    print("OK")
else:
    print("Failed")

OK


In [442]:
# a) How many days are recorded in the dataset?
query_a = "SELECT COUNT(*) AS total_days FROM austin_weather;"
total_days = pd.read_sql(query_a, connection)
print(total_days)

   total_days
0        1319


/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/737266126.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_days = pd.read_sql(query_a, connection)


In [444]:
# b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
query_b = """
SELECT Date, TempHighF
FROM austin_weather
ORDER BY TempHighF DESC
LIMIT 1;
"""
highest_temp_day = pd.read_sql(query_b, connection)
print(highest_temp_day)

        Date  TempHighF
0 2017-07-29        107


/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/1611139015.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  highest_temp_day = pd.read_sql(query_b, connection)


In [446]:
# c) What is the average Humidity across all days? (column HumidityAvgPercent)
query_c = "SELECT AVG(HumidityAvgPercent) AS average_humidity FROM austin_weather;"
average_humidity = pd.read_sql(query_c, connection)
print(average_humidity)

   average_humidity
0         66.570887


/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/1467387042.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  average_humidity = pd.read_sql(query_c, connection)


In [448]:
# d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?
query_d = """
SELECT Date, SeaLevelPressureAvgInches, DewPointAvgF
FROM austin_weather
WHERE DewPointAvgF > 28
ORDER BY SeaLevelPressureAvgInches DESC
LIMIT 10;
"""
top_sealevel_days = pd.read_sql(query_d, connection)
print(top_sealevel_days)

        Date SeaLevelPressureAvgInches DewPointAvgF
0 2016-11-19                     30.54           30
1 2016-12-08                     30.49           31
2 2014-12-31                     30.48           29
3 2017-03-03                     30.48           29
4 2016-01-01                     30.48           33
5 2015-01-04                     30.48           30
6 2014-01-23                     30.47           29
7 2014-11-27                     30.45           33
8 2015-12-04                     30.45           38
9 2015-02-23                     30.45           29


/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/1343145796.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  top_sealevel_days = pd.read_sql(query_d, connection)


#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [451]:
# Your code here



# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [455]:
# Load data into DataFrame
query = "SELECT * FROM austin_weather;"
austin = pd.read_sql(query, connection)

/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/2642412113.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  austin = pd.read_sql(query, connection)


In [461]:
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

In [463]:
austin.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
HumidityLowPercent                    object
SeaLevelPressureHighInches            object
SeaLevelPressureAvgInches             object
SeaLevelPressureLowInches             object
VisibilityHighMiles                   object
VisibilityAvgMiles                    object
VisibilityLowMiles                    object
WindHighMPH                           object
WindAvgMPH                            object
WindGustMPH                           object
PrecipitationSumInches                object
Events                                object
dtype: object

In [465]:
austin.describe()

Date    TempHighF     TempAvgF     TempLowF
count                 1319  1319.000000  1319.000000  1319.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.766523    14.045904    14.190648

In [467]:
austin.shape

(1319, 21)

In [469]:
# Check unique values in problematic columns
for col in ['DewPointHighF', 'HumidityHighPercent', 'SeaLevelPressureHighInches', 'VisibilityHighMiles', 'WindHighMPH', 'PrecipitationSumInches']:
    print(f"Unique values in {col}:")
    print(austin[col].unique())
    print("\n")

Unique values in DewPointHighF:
['67' '43' '31' '36' '44' '39' '41' '49' '51' '37' '55' '54' '15' '29'
 '60' '64' '58' '28' '42' '22' '38' '45' '19' '59' '62' '40' '17' '53'
 '34' '48' '52' '61' '63' '33' '57' '65' '46' '32' '50' '27' '56' '47'
 '68' '69' '66' '70' '73' '71' '76' '72' '75' '74' '-' '77' '23' '35' '30'
 '79' '78' '80' '26' '18' '25' '13']


Unique values in HumidityHighPercent:
['93' '76' '89' '86' '79' '83' '92' '75' '87' '50' '68' '84' '70' '49'
 '55' '45' '59' '64' '71' '78' '81' '85' '66' '69' '63' '80' '82' '58'
 '96' '90' '67' '73' '56' '53' '100' '72' '97' '60' '77' '42' '51' '44'
 '37' '94' '91' '95' '-' '88' '74' '62' '61' '65' '57' '54' '52' '48' '47'
 '43']


Unique values in SeaLevelPressureHighInches:
['29.86' '30.41' '30.56' '30.47' '30.46' '30.33' '30.31' '30.44' '30.49'
 '30.25' '30.48' '30.12' '30.8' '30.66' '30.18' '29.97' '30.16' '30.34'
 '30.21' '30.37' '30.26' '30.35' '30.43' '30.23' '30.73' '30.76' '30.38'
 '30' '30.28' '30.51' '30.52' '29.96' '29.

In [471]:
# Replace "T" with 0 in PrecipitationSumInches and "-" with 0
austin['PrecipitationSumInches'].replace("T", np.nan, inplace=True)
austin.replace("-", np.nan, inplace=True)

/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/499623588.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  austin['PrecipitationSumInches'].replace("T", np.nan, inplace=True)


In [473]:
# Convert non-numeric values
numeric_columns = [
    'DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 
    'HumidityHighPercent', 'HumidityAvgPercent', 'HumidityLowPercent', 
    'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches',
    'VisibilityHighMiles', 'VisibilityAvgMiles', 'VisibilityLowMiles',
    'WindHighMPH', 'WindAvgMPH', 'WindGustMPH', 'PrecipitationSumInches'
]
austin[numeric_columns] = austin[numeric_columns].apply(pd.to_numeric, errors='coerce')

In [475]:
# Replace nan values with median of each column
for col in numeric_columns:
    median_value = austin[col].median() 
    austin[col].fillna(median_value, inplace=True)  

/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/1091676097.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  austin[col].fillna(median_value, inplace=True)


In [477]:
austin.dropna(subset=['Date'], inplace=True)

In [479]:
print("Data types of each column:")
austin.dtypes

Data types of each column:


Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object

In [481]:
print("Descriptive statistics of numeric variables:")
austin.describe()

Descriptive statistics of numeric variables:


Date    TempHighF     TempAvgF     TempLowF  \
count                 1319  1319.000000  1319.000000  1319.000000   
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957   
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000   
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000   
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000   
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000   
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000   
std                    NaN    14.766523    14.045904    14.190648   

       DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
count    1319.000000   1319.000000   1319.000000          1319.000000   
mean       61.539045     56.655800     50.960576            87.875663   
min        13.000000      8.000000      2.000000            37.000000   
25%        53.000000     46.500000     38.000000            85.000000   
50%        66.000000     61.000000     56.000000            90.000000   
75%        73.000000     69.000000     65.000000            94.000000   
max        80.000000     76.000000     75.000000           100.000000   
std        13.554245     14.830139     16.148635            11.030569   

       HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
count         1319.000000         1319.000000                  1319.00000   
mean            66.672479           44.957544                    30.11213   
min             27.000000           10.000000                    29.63000   
25%             59.000000           33.000000                    29.99000   
50%             67.000000           44.000000                    30.08000   
75%             74.000000           55.000000                    30.20500   
max             97.000000           93.000000                    30.83000   
std             12.454180           16.961789                     0.17995   

       SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
count                1319.000000                1319.000000   
mean                   30.022851                  29.931842   
min                    29.550000                  29.410000   
25%                    29.910000                  29.820000   
50%                    30.000000                  29.910000   
75%                    30.100000                  30.020000   
max                    30.740000                  30.610000   
std                     0.171953                   0.171951   

       VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
count          1319.000000         1319.000000         1319.000000   
mean              9.991660            9.169826            6.859742   
min               5.000000            2.000000            0.000000   
25%              10.000000            9.000000            3.000000   
50%              10.000000           10.000000            9.000000   
75%              10.000000           10.000000           10.000000   
max              10.000000           10.000000           10.000000   
std               0.162745            1.454405            3.672381   

       WindHighMPH   WindAvgMPH  WindGustMPH  PrecipitationSumInches  
count  1319.000000  1319.000000  1319.000000             1319.000000  
mean     13.231236     5.008340    21.372252                0.114397  
min       6.000000     1.000000     9.000000                0.000000  
25%      10.000000     3.000000    17.000000                0.000000  
50%      13.000000     5.000000    21.000000                0.000000  
75%      15.000000     6.000000    25.000000                0.000000  
max      29.000000    12.000000    57.000000                5.200000  
std       3.422508     2.084866     5.866770                0.428175

In [483]:
print("First five rows of the dataset:")
austin.head()

First five rows of the dataset:


Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0  ...                      29.68                      29.59   
1  ...                      30.13                      29.87   
2  ...                      30.49                      30.41   
3  ...                      30.45                      30.30   
4  ...                      30.33                      30.27   

   VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
0                 10.0                 7.0                 2.0         20.0   
1                 10.0                10.0                 5.0         16.0   
2                 10.0                10.0                10.0          8.0   
3                 10.0                10.0                 7.0         12.0   
4                 10.0                10.0                 7.0         10.0   

   WindAvgMPH  WindGustMPH  PrecipitationSumInches               Events  
0         4.0         31.0                    0.46  Rain , Thunderstorm  
1         6.0         25.0                    0.00                       
2         3.0         12.0                    0.00                       
3         4.0         20.0                    0.00                       
4         2.0         16.0                    0.00                       

[5 rows x 21 columns]

In [485]:
categorical_columns = austin.select_dtypes(include=['object']).columns
print("Categorical variables:")
print(categorical_columns)

# Display a sample of the categorical columns
print("\nSample data for categorical variables:")
print(austin[categorical_columns].head())

Categorical variables:
Index(['Events'], dtype='object')

Sample data for categorical variables:
                Events
0  Rain , Thunderstorm
1                     
2                     
3                     
4                     


#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


1. The dataset shows significant temperature variation, with high temperatures ranging from 32°F to 107°F, while low temperatures range from 19°F to 81°F. The mean high temperature is around 80.9°F, while the average temperature (mean of TempAvgF) is approximately 70.6°F. This wide range indicates seasonal temperature fluctuations typical of a place with both hot summers and cooler winters, reflecting Austin’s climate. The standard deviations for TempHighF, TempAvgF, and TempLowF (around 14°F) further confirm substantial daily and seasonal variability.
2. HumidityHighPercent reaches 100%, with a median around 90% and an average of approximately 87.9%, indicating frequent high humidity. The variability, as seen in the standard deviation, is lower than temperature fluctuations, suggesting that high humidity is more consistent across the dataset. Dew point values range from 2°F to 80°F, with an average of around 56.7°F, supporting the observation of generally high humidity. High dew points combined with high temperatures suggest periods of warm, humid conditions, typical of Austin’s summer climate.
3. PrecipitationSumInches has a mean of 0.114 inches, but the minimum is 0, with the 25th, 50th, and 75th percentiles also at 0, indicating that most days have no precipitation, while occasional days bring significant rainfall (up to 5.2 inches). The WindHighMPH variable shows a maximum gust speed of 57 MPH, with the mean being around 13.2 MPH. This highlights occasional windy days, but average conditions are calm, with WindAvgMPH at 5 MPH. These precipitation and wind patterns reflect the occasional occurrence of severe weather, such as thunderstorms, which aligns with the "Events" column listing rain and thunderstorms on some days.

#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [491]:
# Display unique values in the DewPointAvgF column
unique_dewpoint = austin['DewPointAvgF'].unique()
print("Unique values in DewPointAvgF:")
unique_dewpoint


Unique values in DewPointAvgF:


array([49., 36., 27., 28., 40., 39., 41., 26., 42., 22., 48., 32.,  8.,
       11., 45., 55., 61., 37., 47., 25., 23., 20., 33., 30., 29., 17.,
       14., 13., 54., 59., 15., 24., 34., 35., 57., 50., 53., 60., 46.,
       56., 51., 31., 38., 62., 43., 63., 64., 67., 66., 58., 70., 68.,
       65., 69., 71., 72., 73., 74., 21., 44., 52., 12., 75., 76., 18.])

- All numeric values
- Float data type
- Consistency values: ranges from 8.0°F to 76.0°F, which seems reasonable for average dew points. No extreme outliers or negative values are present

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [495]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [497]:
# Convert each column 
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

# Verify the conversion by checking data types
print("Data Types after Conversion:")
austin[wrong_type_columns].dtypes

Data Types after Conversion:


DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
dtype: object

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [500]:
object_columns = austin.select_dtypes(include='object').columns
print("\nColumns with object data type:", object_columns)


Columns with object data type: Index(['Events'], dtype='object')


# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [504]:
missing_values = austin[df.isnull().any(axis=1)]
print(missing_values)

Empty DataFrame
Columns: [Date, TempHighF, TempAvgF, TempLowF, DewPointHighF, DewPointAvgF, DewPointLowF, HumidityHighPercent, HumidityAvgPercent, HumidityLowPercent, SeaLevelPressureHighInches, SeaLevelPressureAvgInches, SeaLevelPressureLowInches, VisibilityHighMiles, VisibilityAvgMiles, VisibilityLowMiles, WindHighMPH, WindAvgMPH, WindGustMPH, PrecipitationSumInches, Events]
Index: []

[0 rows x 21 columns]


There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [508]:
# Count the number of rows in the original df
total_rows = austin.shape[0]
print("Total number of rows in the original dataset (austin):", total_rows)

Total number of rows in the original dataset (austin): 1319


In [510]:
# Count the number of rows with missing values
missing_rows = missing_values.shape[0]
print("Total number of rows with missing values:", missing_rows)

Total number of rows with missing values: 0


#### Calculate the ratio of missing rows to total rows

In [513]:
missing_ratio = missing_rows / total_rows
print("Ratio of missing rows to total rows:", missing_ratio)

Ratio of missing rows to total rows: 0.0


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [516]:
# Find the number of missing rows per column
missing_per_column = austin.isna().sum()

# Display the number of missing values in each column
print("Number of missing values per column:")
print(missing_per_column)

Number of missing values per column:
Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [519]:
# Calculate the number of missing values in the PrecipitationSumInches column
missing_precipitation = austin['PrecipitationSumInches'].isna().sum()

# Calculate the ratio of missing values in PrecipitationSumInches to total rows
missing_precipitation_ratio = missing_precipitation / total_rows
print("Ratio of missing values in PrecipitationSumInches to total rows:", missing_precipitation_ratio)

Ratio of missing values in PrecipitationSumInches to total rows: 0.0


Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [522]:
# Drop the PrecipitationSumInches column from the dataset
austin.drop('PrecipitationSumInches', axis=1, inplace=True)

# Verify that the column has been removed
print("Columns after dropping PrecipitationSumInches:")
austin.columns

Columns after dropping PrecipitationSumInches:


Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events'],
      dtype='object')

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [525]:
# Your code here
# Perform linear interpolation on the dataframe
austin_fixed = austin.interpolate(inplace=False)

# Display both dataframes for comparison
print("Original DataFrame with Missing Data:")
print(austin.head())  

print("\nInterpolated DataFrame:")
print(austin_fixed.head()) 

Original DataFrame with Missing Data:
        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighI

/var/folders/0b/0nt_hmld68j1x9vbn6980snc0000gn/T/ipykernel_5283/419731288.py:3: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  austin_fixed = austin.interpolate(inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [528]:
# Your code here
austin_fixed.isnull().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

In [530]:
austin.isnull().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [534]:
# Your code here:
# Evaluate the contents of the Events column
events_counts = austin['Events'].value_counts()

# Display the unique values and their counts
print("Contents of the 'Events' column:")
print(events_counts)

Contents of the 'Events' column:
Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64


Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [537]:
# Your answer:
# Calculate the maximum number of events in a single day
max_events = austin['Events'].dropna().apply(lambda x: len(x.split(','))).max()

print("The largest number of events that happened in a single day is:", max_events)

The largest number of events that happened in a single day is: 3


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [543]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     0            0  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [547]:
# Your code here
# Populate the dummy columns based on the Events column
for event in event_list:
    austin_fixed[event] = austin_fixed['Events'].str.contains(event, na=False).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [549]:
# Your code here
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     1            1  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [551]:
# Your code here
austin_fixed = austin_fixed.drop(columns=['Events'])

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [555]:
# Your code here
from datetime import datetime

# Convert the Date column to datetime objects and then to ordinal numbers
austin_fixed['Date'] = austin_fixed['Date'].apply(lambda x: x.toordinal() if isinstance(x, (datetime, pd.Timestamp)) else datetime.strptime(x, '%Y-%m-%d').toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [558]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [560]:
# Your code here:
# Assign predictor variables to X
X = austin_fixed[
    [
        'TempHighF',
        'TempLowF',
        'DewPointHighF',
        'DewPointAvgF',
        'DewPointLowF',
        'HumidityHighPercent',
        'HumidityAvgPercent',
        'HumidityLowPercent',
        'SeaLevelPressureHighInches',
        'SeaLevelPressureAvgInches',
        'SeaLevelPressureLowInches',
        'VisibilityHighMiles',
        'VisibilityAvgMiles',
        'VisibilityLowMiles',
        'WindHighMPH',
        'WindAvgMPH',
        'WindGustMPH',
        'Snow',
        'Fog',
        'Rain',
        'Thunderstorm',
    ]
]

# Assign response variable to y
y = austin_fixed['TempAvgF']

# Display the first few rows of X and y to verify
print("Predictor variables (X):")
print(X.head())

print("\nResponse variable (y):")
print(y.head())

Predictor variables (X):
   TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0         74        45           67.0          49.0          43.0   
1         56        39           43.0          36.0          28.0   
2         58        32           31.0          27.0          23.0   
3         61        31           36.0          28.0          21.0   
4         58        41           44.0          40.0          36.0   

   HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                 93.0                75.0                57.0   
1                 93.0                68.0                43.0   
2                 76.0                52.0                27.0   
3                 89.0                56.0                22.0   
4                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  ...  \
0                       29.86                      29.68  ...   
1                       30.41    

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [562]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [564]:
#Your code here:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the sizes of the resulting datasets
print("Training set:")
print(f"X_train shape: {X_train.shape}")
print(f"y_train shape: {y_train.shape}")

print("\nTest set:")
print(f"X_test shape: {X_test.shape}")
print(f"y_test shape: {y_test.shape}")

Training set:
X_train shape: (1055, 21)
y_train shape: (1055,)

Test set:
X_test shape: (264, 21)
y_test shape: (264,)


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [382]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [384]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [386]:
# Your code here:
